### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by JeremyMonarch**
Add xUnit tests

### CODE REVIEW

File: Preparation/MyStack/MyStack.csproj [hunk 1]
```diff
@@ -1,8 +1,16 @@
-<Project Sdk="Microsoft.NET.Sdk">
```
Comment (Nazakato) [inline Preparation/MyStack/MyStack.csproj]: You don't need changes in next files:
- Preparation/MyStack/MyStack.csproj
- Preparation/Preparation/Spiral_Array.csproj
- Preparation/Sand_Clock/Sand_Clock.csproj

File: Preparation/Preperation.Tests/MyStackInterTest.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Preperation.Tests
+{
+    public class MyStackInterTest
+    {
+
+    }
```
Comment (Nazakato) [inline Preparation/Preperation.Tests/MyStackInterTest.cs]: It seems you have created 2 separate test projects. Please leave single, the Preparation.Tests. Move into MyStackInterTest.cs all the tests

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+using MyStack;
+using Xunit;
+
+namespace Preperation.Tests
+{
+    public class MyStackInterTest
+    {
+        [Fact]
+        public void PushTest()
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs]: Use naming convention that describes the intention of the unit test. E.g. one of the widely used: MethodName_DoesWhat_WhenTheseConditions. In this case, it would be named like Push_AddsOneElementToTheCollection

And inside the test I would check that after pushing an element to an empty collection, the collection has single element.
Using linq:
Assert.Equal(1, myStack.Count())
Assert.Equal(1, myStack.Single())

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            MyStack<int> myStack = new MyStack<int>();
+
+            myStack.Push(1);
+
+            Assert.Equal(1, myStack.Peek());
+        }
+        [Fact]
+        public void PeekTest()
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs]: I would test that Peek takes the last pushed element, but doesn't change the count of the elements.

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            myStack.Push(1);
+            myStack.Push(2);
+            myStack.Push(3);
+
+            int result = myStack.Peek();
+
+            Assert.Equal(3, result);
+        }
+        [Fact]
+        public void PopTest()
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs]: As Peek, but changes the collection

Comment (Nazakato) [reply Preparation/TestProject1/MyStackInterTest.cs]: Another case you need to cover with tests - pop when collectoin is already empty. Excpetion should be thrown as far as I remember

File: Preparation/TestProject1/MyStackInterTest.cs (hunk 3 above)
Comment (Nazakato) [reply Preparation/TestProject1/MyStackInterTest.cs]: Using LINQ is better than using you're own method for testing. Because with this test you're not testing your methods in isolation one from another.

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+namespace Preperation.Tests
+{
+    public class MyStackInterTest
+    {
+        [Fact]
+        public void Push_AddsOneElementToTheCollection()
+        {
+            MyStack<int> myStack = new MyStack<int>();
+
+            myStack.Push(1);
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs:16]: It is enough to push only one element in this test

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            myStack.Push(4);
+
+            Assert.Equal(4, myStack.Count());
+            Assert.Equal(4, myStack.Last());
+        }
+        [Fact]
+        public void Peek_ReturnsLastPushedElementsInTheCollectionTest()
+        {
+            MyStack<int> myStack = new MyStack<int>();
+            myStack.Push(1);
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs:37]: Two pushes are enough

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            myStack.Push(20);
+            myStack.Push(30);
+
+            int result = myStack.Peek();
+
+            Assert.Equal(30, result);
+            Assert.Equal(3, myStack.Count());
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs]: This should be a separate test that specifically checks that stack elements count is not changed during peek. Similar should be applied for checking Pop work

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            Assert.Equal(2, myStack.Count());
+        }
+        [Fact]
+        public void Pop_TryingRemoveElementInEmptyCollection()
+        {
+            MyStack<int> myStack = new MyStack<int>();
+
+            myStack.Push(1);
+
+            int result = myStack.Pop();
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs:72]: This test should check that emptyCollection.Pop/Peek throws. You just initialize and do not push any elements - no need.

Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs:72]: This test should check that emptyCollection.Pop/Peek throws. You just initialize and do not push any elements - no need.

File: Preparation/TestProject1/MyStackInterTest.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            myStack.Push(2);
+            myStack.Push(3);
+
+            int result = myStack.Pop();
+
+            Assert.Equal(3, result);
+            Assert.Equal(2, myStack.Count());
+        }
+        [Fact]
+        public void Pop_TryingRemoveElementInEmptyCollection()
```
Comment (Nazakato) [inline Preparation/TestProject1/MyStackInterTest.cs]: Pop_ThrowsInvalidOperqtionException_WhenStackIsEmpty
Use single naming convention everywhere

File: Preparation/TestProject1/MyStackInterTest.cs (hunk 9 above)
Comment (Nazakato) [reply Preparation/TestProject1/MyStackInterTest.cs:72]: Just as you do in the next test (isemptycheck)

**Review by Nazakato — CHANGES_REQUESTED** [review]
1 Unnecessary files still have changes. If you need help with git soft reset, I can help, just ping
2 Stick to single readable naming convention. General advise - try to read your code as a set of statements/instructions in human readable language. Naming matters.

File: Preparation/TestProject1/MyStackInterTest.cs (hunk 8 above)
Comment (JeremyMonarch) [reply Preparation/TestProject1/MyStackInterTest.cs]: So I need to write additional test-methods, which check that count is correct work in Peek() and Pop(), right?

Comment (Nazakato) [reply Preparation/TestProject1/MyStackInterTest.cs]: Instead of this one `Peek_ReturnsLastPushedElementsInTheCollectionTest` you should have two
`Peek_ReturnsLastPushedElement` and `Peek_DoesNotChangeCount`

**Review by Nazakato — APPROVED** [review]
